### PR TITLE
allow the beaconChainController to skip forward slots if it falls behind

### PR DESCRIPTION
**Pull request by rolfyone**
- added a getSlotStartTime function to calculate the time that a slot should start

- added regression tests for getSlotStartTime

- added a event log event for when the nodeSlot drifts, so that we can see it in the log, as it will likely mean that errors occur slightly after, especially if it is multiple epochs between them.

- removed Thread.sleep() logic previously used for slot processing logic.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

Addresses #1942
## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: logging/src/main/java/tech/pegasys/teku/logging/EventLogger.java [hunk 1]
```diff
@@ -56,6 +56,14 @@ public void epochEvent(
     info(epochEventLog, Color.GREEN);
   }
 
+  public void nodeSlotDriftEvent(final UnsignedLong oldSlot, final UnsignedLong newSlot) {
+    final String driftEventLog =
+        String.format(
+            "Drift Event *** Current slot: %s, previous slot: %s",
```
Comment (ajsutton) [inline logging/src/main/java/tech/pegasys/teku/logging/EventLogger.java]: Drift Event probably isn't the right wording here.  Probably just "Skipped Slots"?  I don't think we always need Event in there and possibly should just remove them all as redundant.

File: services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java [hunk 2]
```diff
@@ -513,12 +520,46 @@ public void onTick(Date date) {
...
+    }
+
+    final UnsignedLong genesisTime = recentChainData.getGenesisTime();
+    final UnsignedLong calculatedSlot =
+        currentTime.compareTo(genesisTime) < 0
+            ? ZERO
+            : ForkChoiceUtil.getCurrentSlot(currentTime, genesisTime);
+    if (currentTime.compareTo(genesisTime) < 0) {
+      return;
+    }
```
Comment (ajsutton) [inline services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java]: We bail out if we're prior to geneesis time so can just do that check before the calculatedSlot calculation.

File: services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java [hunk 3]
```diff
@@ -513,12 +520,46 @@ public void onTick(Date date) {
...
+      onTickSlotAttestation = nodeSlot.getValue();
+      processSlotAttestation(epoch);
+    }
+    if ((onTickSlotAggregate == null || calculatedSlot.compareTo(onTickSlotAggregate) > 0)
+        && currentTime.compareTo(
+                nodeSlotStartTime.plus(UnsignedLong.valueOf(oneThirdSlotSeconds * 2)))
+            >= 0) {
+      onTickSlotAggregate = nodeSlot.getValue();
+      processSlotAggregate();
     }
```
Comment (ajsutton) [inline services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java]: I think I'm following how this all works but it's not easy. Will  take another look in the morning - I suspect that extracting some private methods would help a lot to give names to the various comparisons being done.

File: logging/src/main/java/tech/pegasys/teku/logging/EventLogger.java (hunk 1 above)
Comment (rolfyone) [reply logging/src/main/java/tech/pegasys/teku/logging/EventLogger.java]: skipped slots is too long... unless i move all the messages ?

File: services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java (hunk 3 above)
Comment (rolfyone) [reply services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java]: yep that's why i was considering refactoring, i don't really like it overly, but the next step would be to refactor...

File: services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java (hunk 2 above)
Comment (rolfyone) [reply services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java]: fair point.

File: logging/src/main/java/tech/pegasys/teku/logging/EventLogger.java (hunk 1 above)
Comment (rolfyone) [reply logging/src/main/java/tech/pegasys/teku/logging/EventLogger.java]: I replaced with 'Miss Slots' - thoughts?

**Review by ajsutton — APPROVED** [review]
Decided to do a bigger refactoring in a follow up PR so will merge this as-is.  The logic is correct just a little hard to read.

**Pull request merged**